### PR TITLE
Periodic dispatcher

### PR DESCRIPTION
**Pull request by dadgar**
This PR adds support for running periodic jobs. It provides an at-most once run guarantee even across leadership transitions.

A follow up PR is needed to force run a periodic job.

### CODE REVIEW

**Comment by c4milo** [comment]
Awesome work @dadgar, I'm going to test this PR as soon as possible.

**Comment by ranjib** [comment]
When periodic jobs are created (i.e. nomad run periodic.hcl), the CLI does not show evaluation details similar to batch or service jobs. Is it expected? Similarly status for jobs also show <none>, i was expecting that to show either next trigger time or something else, and probably previous allocations etc.

**Comment by dadgar** [comment]
@ranjib: It is expected that the CLI doesn't show eval details on periodic jobs. This is because periodic jobs never have an evaluation associated with them. They will spawn child jobs that will have evals but they themselves never do.

As for `nomad status`, that can be improved. This PR was just getting too big.

**Comment by ranjib** [comment]
@dadga :tada:  One more thing, system hobs cant be periodic? like we cant use nomad to run chef periodically on all hosts?
FYI i took this branch for a spin under the [systemd driver] and its working well.

**Comment by dadgar** [comment]
@ranjib: Currently (and for the foreseeable future) there will not be periodic system jobs. You could probably create the behavior by registering a system job and on the interval you need to run chef, do an update to the job that would force Nomad to rerun the job (changing args would do it).

**Comment by c4milo** [comment]
I started testing this branch using the `exec` driver, but I've been noticing something weird with the allocations. For some reason once the task finishes, either because it failed running or it died, the allocation doesn't seem to be released and after a while I start getting messages about resources being exhausted. I'm running locally using Vagrant and the VM is pretty much available for any workload.

**Comment by dadgar** [comment]
@c4milo: Yeah that isn't actually a problem with this PR. It is a bigger problem in general. We aren't freeing up the resources consumed by an alloc even when it is terminal in certain cases. This will be fixed by a PR I am working on now. Should be done soon.

File: nomad/structs/structs.go [hunk 1]
```diff
@@ -916,6 +938,10 @@ func (u *UpdateStrategy) Rolling() bool {
 const (
 	// PeriodicSpecCron is used for a cron spec.
 	PeriodicSpecCron = "cron"
+
+	// PeriodicSpecTest is only used by unit tests. It is a sorted, comma
+	// seperated list of unix timestamps at which to launch.
+	PeriodicSpecTest = "test"
```
Comment (armon) [inline nomad/structs/structs.go]: Given this behavior, maybe it should just be called "fixed" since it runs at fixed times. Otherwise, maybe pick something less likely to be accidentally used, e.g. "_internal_test"

File: nomad/state/schema.go [hunk 2]
```diff
@@ -100,6 +101,45 @@ func jobTableSchema() *memdb.TableSchema {
...
+				Indexer: &memdb.StringFieldIndex{
+					Field:     "ParentID",
+					Lowercase: true,
+				},
+			},
+			"periodic": &memdb.IndexSchema{
+				Name:         "periodic",
+				AllowMissing: false,
+				Unique:       false,
+				Indexer: &memdb.FieldSetIndex{
```
Comment (armon) [inline nomad/state/schema.go]: ConditionalIndexer makes more sense to scan for enabled I'm guessing

File: nomad/state/schema.go [hunk 3]
```diff
@@ -100,6 +101,45 @@ func jobTableSchema() *memdb.TableSchema {
...
+			// Primary index is used for job management
+			// and simple direct lookup. ID is required to be
+			// unique.
+			"id": &memdb.IndexSchema{
+				Name:         "id",
+				AllowMissing: false,
+				Unique:       true,
+				Indexer: &memdb.StringFieldIndex{
+					Field:     "ID",
+					Lowercase: true,
```
Comment (armon) [inline nomad/state/schema.go]: Is there a reason to lower case instead of case matching?

File: nomad/structs/structs.go [hunk 4]
```diff
@@ -963,11 +991,40 @@ func (p *PeriodicConfig) Next(fromTime time.Time) time.Time {
...
+		}
 	}
 
 	return time.Time{}
 }
 
+// PeriodicLaunch tracks the last launch time of a periodic job.
+type PeriodicLaunch struct {
+	ID     string    // ID of the periodic job.
+	Launch time.Time // The last launch time.
```
Comment (armon) [inline nomad/structs/structs.go]: Add the CreateIndex and ModifyIndex

File: nomad/state/state_store.go [hunk 5]
```diff
@@ -372,6 +395,92 @@ func (s *StateStore) JobsByScheduler(schedulerType string) (memdb.ResultIterator
...
+func (s *StateStore) UpsertPeriodicLaunch(index uint64, launch *structs.PeriodicLaunch) error {
+	txn := s.db.Txn(true)
+	defer txn.Abort()
+
+	watcher := watch.NewItems()
+	watcher.Add(watch.Item{Table: "periodic_launch"})
+	watcher.Add(watch.Item{Job: launch.ID})
+
+	// Check if the job already exists
+	if _, err := txn.First("periodic_launch", "id", launch.ID); err != nil {
```
Comment (armon) [inline nomad/state/state_store.go]: Existing entry should be used to populate the CreateIndex

File: nomad/state/state_store.go [hunk 6]
```diff
@@ -372,6 +395,92 @@ func (s *StateStore) JobsByScheduler(schedulerType string) (memdb.ResultIterator
...
+	txn.Commit()
+	return nil
+}
+
+// DeletePeriodicLaunch is used to delete the periodic launch
+func (s *StateStore) DeletePeriodicLaunch(index uint64, jobID string) error {
+	txn := s.db.Txn(true)
+	defer txn.Abort()
+
+	watcher := watch.NewItems()
```
Comment (armon) [inline nomad/state/state_store.go]: May as well move this until after the check for existing, since it may not be needed

File: nomad/periodic.go [hunk 7]
```diff
@@ -0,0 +1,529 @@
...
+
+const (
+	// The string appended to the periodic jobs ID when launching derived
+	// instances of it.
+	JobLaunchSuffix = "/periodic-"
+)
+
+// PeriodicRunner is the interface for tracking and launching periodic jobs at
+// their periodic spec.
+type PeriodicRunner interface {
```
Comment (armon) [inline nomad/periodic.go]: I'd change this around such that there is no `PeriodicRunner` interface, but instead of passing the server into `PeriodicDispatch` you pass that in as an interface. e.g. the `EvalDispatch` interface. That way you can test the PeriodicDispatch more thoroughly by mocking the server itself. There is no other expected implementation of PeriodicRunner

File: nomad/periodic.go [hunk 8]
```diff
@@ -0,0 +1,529 @@
...
+// Start begins the goroutine that creates derived jobs and evals.
+func (p *PeriodicDispatch) Start() {
+	p.l.Lock()
+	p.running = true
+	p.l.Unlock()
+	go p.run()
+}
+
+// Tracked returns the set of tracked job IDs.
+func (p *PeriodicDispatch) Tracked() []structs.Job {
```
Comment (armon) [inline nomad/periodic.go]: This should be a list of pointers instead

File: nomad/periodic.go [hunk 9]
```diff
@@ -0,0 +1,529 @@
...
+
+	// Do nothing if not enabled
+	if !p.enabled {
+		return nil
+	}
+
+	// If we were tracking a job and it has been disabled or made non-periodic remove it.
+	disabled := !job.IsPeriodic() || !job.Periodic.Enabled
+	_, tracked := p.tracked[job.ID]
+	if tracked && disabled {
```
Comment (armon) [inline nomad/periodic.go]: Maybe clearer as:
```
if disabled {
  if tracked {
     p.removeLocked(job.ID)
  }
  return nil
}
```

File: nomad/periodic.go [hunk 10]
```diff
@@ -0,0 +1,529 @@
...
+	p.l.RUnlock()
+
+	var now time.Time
+
+PICK:
+	// If there is nothing wait for an update.
+	p.l.RLock()
+	if p.heap.Length() == 0 {
+		p.l.RUnlock()
+		p.logger.Printf("[DEBUG] nomad.periodic: no periodic jobs; waiting")
```
Comment (armon) [inline nomad/periodic.go]: Verbose and useless log

File: nomad/periodic.go [hunk 11]
```diff
@@ -0,0 +1,529 @@
...
+		default:
+			p.logger.Printf("[ERR] nomad.periodic: failed to determine next periodic job: %v", err)
+			return
+		}
+	}
+
+	launchTime := nextJob.next
+
+	// If there are only invalid times, wait for an update.
+	if launchTime.IsZero() {
```
Comment (armon) [inline nomad/periodic.go]: How could this happen? This seems like a fatal condition

File: nomad/periodic.go [hunk 12]
```diff
@@ -0,0 +1,529 @@
...
+	if !tracked {
+		return fmt.Errorf("can't force run non-tracked job %v", jobID)
+	}
+
+	return p.createEval(job, time.Now())
+}
+
+// run is a long-lived function that waits til a job's periodic spec is met and
+// then creates an evaluation to run the job.
+func (p *PeriodicDispatch) run() {
```
Comment (armon) [inline nomad/periodic.go]: This whole method can be simplified to something like

```
for p.shouldRun() {
    getNextRunOrNil := ....

   select {
   case stop: 
      return
   case nextRun:
       dispatch()
   case update:
      continue
   }
}
```

File: nomad/periodic.go [hunk 13]
```diff
@@ -0,0 +1,529 @@
...
+		p.l.Unlock()
+		now = nowUpdate
+	}
+
+	goto PICK
+}
+
+// createEval instantiates a job based on the passed periodic job and submits an
+// evaluation for it.
+func (p *PeriodicDispatch) createEval(periodicJob *structs.Job, time time.Time) error {
```
Comment (armon) [inline nomad/periodic.go]: This is the logic you want to abstract out into an interface to more easily test both separately

File: nomad/periodic.go [hunk 14]
```diff
@@ -0,0 +1,529 @@
...
+	return nil
+}
+
+// deriveJob instantiates a new job based on the passed periodic job and the
+// launch time.
+// TODO: these jobs need to be marked as GC'able
+func (p *PeriodicDispatch) deriveJob(periodicJob *structs.Job, time time.Time) (
+	derived *structs.Job, err error) {
+
+	// Have to recover in case the job copy panics.
```
Comment (armon) [inline nomad/periodic.go]: Why is this possible? We can do an explicit copy of all the modified fields instead of depending on copy structure if this is an issue.

File: nomad/leader.go [hunk 15]
```diff
@@ -167,6 +176,44 @@ func (s *Server) restoreEvalBroker() error {
...
+	iter, err := s.fsm.State().JobsByPeriodic(true)
+	if err != nil {
+		return fmt.Errorf("failed to get periodic jobs: %v", err)
+	}
+
+	now := time.Now()
+	for i := iter.Next(); i != nil; i = iter.Next() {
+		job := i.(*structs.Job)
+		s.periodicDispatcher.Add(job)
+
```
Comment (armon) [inline nomad/leader.go]: We should probably guard on is the periodic dispatcher enabled

File: nomad/leader.go [hunk 16]
```diff
@@ -167,6 +176,44 @@ func (s *Server) restoreEvalBroker() error {
...
+		launch, err := s.fsm.State().PeriodicLaunchByID(job.ID)
+		if err != nil || launch == nil {
+			return fmt.Errorf("failed to get periodic launch time: %v", err)
+		}
+
+		if !job.Periodic.Next(launch.Launch).Before(now) {
```
Comment (armon) [inline nomad/leader.go]: Can you add some more comments here? This is a really important line but its pretty hard to reason about it especially without any supporting comments

File: nomad/fsm.go [hunk 17]
```diff
@@ -204,6 +207,49 @@ func (n *nomadFSM) applyUpsertJob(buf []byte, index uint64) interface{} {
...
+
+	// If it is periodic, insert it into the periodic runner and record the
+	// time it was inserted.
+	if req.Job.IsPeriodic() {
+		if err := n.periodicRunner.Add(req.Job); err != nil {
+			n.logger.Printf("[ERR] nomad.fsm: PeriodicRunner.Add failed: %v", err)
+			return err
+		}
+
+		// Record the insertion time as a launch.
```
Comment (armon) [inline nomad/fsm.go]: Can you add more comments explaining why this is necessary. I get it, but its not immediately obvious.

File: nomad/fsm.go [hunk 18]
```diff
@@ -204,6 +207,49 @@ func (n *nomadFSM) applyUpsertJob(buf []byte, index uint64) interface{} {
...
+			return nil
+		}
+
+		if parent.IsPeriodic() {
+			t, err := n.periodicRunner.LaunchTime(req.Job.ID)
+			if err != nil {
+				n.logger.Printf("[ERR] nomad.fsm: LaunchTime(%v) failed: %v", req.Job.ID, err)
+				return err
+			}
+			launch := &structs.PeriodicLaunch{parentID, t}
```
Comment (armon) [inline nomad/fsm.go]: Use the named fields

File: nomad/fsm.go [hunk 19]
```diff
@@ -204,6 +207,49 @@ func (n *nomadFSM) applyUpsertJob(buf []byte, index uint64) interface{} {
...
+		if err != nil {
+			n.logger.Printf("[ERR] nomad.fsm: JobByID(%v) lookup for parent failed: %v", parentID, err)
+			return err
+		} else if parent == nil {
+			// The parent has been deregistered.
+			return nil
+		}
+
+		if parent.IsPeriodic() {
+			t, err := n.periodicRunner.LaunchTime(req.Job.ID)
```
Comment (armon) [inline nomad/fsm.go]: Why is this necessary? Why not use the current time?

File: nomad/fsm.go [hunk 20]
```diff
@@ -204,6 +207,49 @@ func (n *nomadFSM) applyUpsertJob(buf []byte, index uint64) interface{} {
 		n.logger.Printf("[ERR] nomad.fsm: UpsertJob failed: %v", err)
 		return err
 	}
+
+	// If it is periodic, insert it into the periodic runner and record the
+	// time it was inserted.
+	if req.Job.IsPeriodic() {
```
Comment (armon) [inline nomad/fsm.go]: If the job is not periodic but was previously, we need to remove it from the periodic runner

File: nomad/fsm.go (hunk 19 above)
Comment (dadgar) [inline nomad/fsm.go]: It is less precise but given that the difference in time is negligible and the added complexity I agree. Will be going through all the comments soon

**Comment by c4milo** [comment]
I'm curious about why this cannot be implemented at the task level. It seems we could leverage more existing Nomad features if implemented this way as well, such as colocating cron tasks executed at different frequencies that also share state, for instance.

File: nomad/state/schema.go (hunk 3 above)
Comment (dadgar) [inline nomad/state/schema.go]: Not really. The `id` here is the job.ID and the index on the job table does lowercase as well. If we should change it, it should be in both.

File: nomad/structs/structs.go (hunk 4 above)
Comment (dadgar) [inline nomad/structs/structs.go]: Done

File: nomad/state/state_store.go (hunk 6 above)
Comment (dadgar) [inline nomad/state/state_store.go]: Done and did it for the other `Delete` methods

File: nomad/periodic.go (hunk 14 above)
Comment (dadgar) [inline nomad/periodic.go]: This doesn't ever happen in practice. It is just protection. There is also a test to let us catch if CopyStructure ever does panic: https://github.com/hashicorp/nomad/pull/540/files#diff-d8207d79ddf876021a4e0225297fa6b9R96

File: nomad/periodic.go (hunk 11 above)
Comment (dadgar) [inline nomad/periodic.go]: Explained offline but will put it here in case anyone else is interested. You can have valid specs that will have no subsequent launches. Such as a cron spec that says launch everyday in January of 2015.

File: nomad/leader.go (hunk 15 above)
Comment (dadgar) [inline nomad/leader.go]: How is it any different than restoreEvalBroker? The periodicDispatcher no-ops when it is not enabled as well

**Comment by ketzacoatl** [comment]
> @ranjib: Currently (and for the foreseeable future) there will not be periodic system jobs. You could probably create the behavior by registering a system job and on the interval you need to run chef, do an update to the job that would force Nomad to rerun the job (changing args would do it).

@dadgar, forgive for me for not immediately seeing the logic in this - what is the reasoning here? As I read https://www.nomadproject.io/docs/jobspec/schedulers.html, a "System" cron task is _exactly_ how I would expect to use the cron-type of scheduling.

**Comment by dadgar** [comment]
@ketzacoatl: The use cases and the subsequent architecture of these two features is prohibitive of using periodic with system jobs.

System jobs are registered and triggered to be evaluated every time node state changes. A periodic job works by spawning a child job and evaluating it every time  the periodic spec is met (every hour/once a day etc). So if you marked a system job as periodic, every time the periodic spec is met, a new system job would be generated and this would lead to undesirable behavior.

The main use case for periodic is to run your batch jobs at a given interval and the main use case for system jobs is to register long lived services.

 To achieve @ranjib goal, you could do what I described earlier or better yet, have your submitted job be a wrapper of chef that launches it periodically.

**Comment by ketzacoatl** [comment]
I'm sure it makes sense from the internals perspective.. but from that of a newb user, having a "periodic" job that is also a "system" job, intended to run as such (except maybe not when nodes change state, make it optional?), makes a whole lot of sense. I think I can force that to make sense, but it feels a whole lot like a square peg going into a round hole.

Thanks for your hard work on this, I will look forward to understanding how to make the best out of these new features!

**Comment by c4milo** [comment]
@dadgar will this PR land in 0.3.0?

**Comment by c4milo** [comment]
@dadgar, another question is what would happen if a periodic job is still running when a new execution kicks in. Does at-most once mean multiple jobs of the same type could be running at the same time?

**Comment by dadgar** [comment]
@c4milo: Yes it will land in 0.3.0. Yes currently multiple jobs can be spawned and running at the same time. A follow up PR will add a config to the `periodic` block (and the necessary logic) to disallow overlapping runs.

**Comment by c4milo** [comment]
>  Yes it will land in 0.3.0.

Excellent! 

> Yes currently multiple jobs can be spawned and running at the same time. A follow up PR will add a config to the periodic block (and the necessary logic) to disallow overlapping runs.

While the PR arrives, I think it will make the current work much more safer to use if we disallow concurrent runs by default.

File: nomad/fsm.go [hunk 21]
```diff
@@ -214,10 +262,29 @@ func (n *nomadFSM) applyDeregisterJob(buf []byte, index uint64) interface{} {
...
+		n.logger.Printf("[ERR] nomad.fsm: DeleteJob failed: %v", err)
+		return err
+	}
+
 	if err := n.state.DeleteJob(index, req.JobID); err != nil {
 		n.logger.Printf("[ERR] nomad.fsm: DeleteJob failed: %v", err)
 		return err
 	}
+
+	if job.IsPeriodic() {
```
Comment (armon) [inline nomad/fsm.go]: We should remove this guard, and always clear out from periodicDispatcher and the launch table, since we may have taken a periodic job and made it non-periodic

File: nomad/state/schema.go [hunk 22]
```diff
@@ -109,6 +110,23 @@ func jobTableSchema() *memdb.TableSchema {
 					Conditional: jobIsGCable,
 				},
 			},
+			"parent": &memdb.IndexSchema{
```
Comment (armon) [inline nomad/state/schema.go]: Remove index

File: nomad/structs/structs.go [hunk 23]
```diff
@@ -944,6 +970,8 @@ func (p *PeriodicConfig) Validate() error {
 		if _, err := cronexpr.Parse(p.Spec); err != nil {
 			return fmt.Errorf("Invalid cron spec %q: %v", p.Spec, err)
 		}
+	case PeriodicSpecTest:
+		// No-op
 	default:
 		return fmt.Errorf("Unknown specification type %q", p.SpecType)
```
Comment (armon) [inline nomad/structs/structs.go]: Make this more clear with "Unknown periodic specification type"

File: nomad/fsm.go [hunk 24]
```diff
@@ -204,6 +207,51 @@ func (n *nomadFSM) applyUpsertJob(buf []byte, index uint64) interface{} {
...
+
+	// If it is periodic, record the time it was inserted. This is necessary for
+	// recovering during leader election. It is possible that from the time it
+	// is added to when it was suppose to launch, leader election occurs and the
+	// job was not launched. In this case, we use the insertion time to
+	// determine if a launch was missed.
+	if req.Job.IsPeriodic() {
+		// Record the insertion time as a launch. We overload the launch table
+		// such that the first entry is the insertion time.
+		launch := &structs.PeriodicLaunch{ID: req.Job.ID, Launch: time.Now()}
```
Comment (armon) [inline nomad/fsm.go]: This should only be updated if there is no existing launch entry

File: nomad/fsm.go [hunk 25]
```diff
@@ -214,10 +262,29 @@ func (n *nomadFSM) applyDeregisterJob(buf []byte, index uint64) interface{} {
 		panic(fmt.Errorf("failed to decode request: %v", err))
 	}
 
+	job, err := n.state.JobByID(req.JobID)
```
Comment (armon) [inline nomad/fsm.go]: Don't need to lookup job if we are not checking for periodic

**Comment by c4milo** [comment]
:dancers:

**Comment by c4milo** [comment]
This is a beautiful christmas gift :christmas_tree:
